cql3/statements/select_statement.cc

```diff
@@ -542,10 +542,14 @@ indexed_table_select_statement::do_execute_base_query(
                 if (old_paging_state->get_clustering_key() && _schema->clustering_key_size() > 0) {
                     auto base_ck = generate_base_key_from_index_pk<clustering_key>(old_paging_state->get_partition_key(),
                             old_paging_state->get_clustering_key(), *_schema, *_view_schema);
-                    command->slice.set_range(*_schema, base_pk,
-                            std::vector<query::clustering_range>{query::clustering_range::make_starting_with(range_bound<clustering_key>(base_ck, false))});
+
+                    auto row_ranges = command->slice.default_row_ranges();
+                    query::trim_clustering_row_ranges_to(*_schema, row_ranges, base_ck, false);
+                    command->slice.set_range(*_schema, base_pk, row_ranges);
                 } else {
-                    command->slice.set_range(*_schema, base_pk, std::vector<query::clustering_range>{query::clustering_range::make_open_ended_both_sides()});
+                    // If there are none _specific_ranges, the entire partition range
+                    // will be read.
```

Comment:
I'm confused by this comment.  Why, specifically, do we not need to set_range on command->slice when CK is involved in paging?  Does some other code set it for us?  What do specific_ranges have to do with anything?

Reply:
This base_ck is computed from old_paging_state, @dekimir. As far as I understand, paging state for index has a clustering key of a base table encoded inside it. This code extracts last returned base table ck and uses it to return only stuff after this last already returned ck.

Reply:
But the comment doesn't say that.  It doesn't mention CK at all.

The reader will be wondering why we're asserting here instead of calling `clear_ranges()`; this comment should explain that.

Reply:
So the explanation of this is quite complex to succinctly describe:

Before the change, the code was: 
```cpp
command->slice.set_range(*_schema, base_pk, 
   std::vector<query::clustering_range>{query::clustering_range::make_open_ended_both_sides()});
```
To fix the bug I found, we would have to intersect this range with `_row_ranges`, the same way as in non-else branch. However, intersection of `_row_ranges` and `clustering_range::make_open_ended_both_sides` is just `_row_ranges`. This would result in such code:
```cpp
command->slice.set_range(*_schema, base_pk, command->slice.default_row_ranges());
```
But because `_specific_ranges` (those are set by `set_range`) are `nullptr` as created in `prepare_command_for_base_query`, this `set_range` is unnecessary, because if no `_specific_ranges` are set, `_row_ranges` are used (returned by `default_row_ranges()`, more info: https://github.com/scylladb/scylla/issues/7362). 

I added this assert to make sure that if someone changes `prepare_command_for_base_query` to add `_specific_ranges`, they will get assert failure in unit tests, because the necessary precondition is not true anymore.

Reply:
@dekimir In the amended commit, I changed this comment to more thoroughly explain this and overarching context.

Reply:
> To fix the bug I found, we would have to intersect this range with `_row_ranges`, the same way as in non-else branch

Then why don't we do precisely that?  It doesn't matter that it currently happens to amount to a no-op, it's future-proof correct code.

> I added this assert to make sure that if someone changes `prepare_command_for_base_query`

Wouldn't it be better if the new code still worked even in this scenario?  Why burden the future developer with a complicated explanation they'll likely struggle with, instead of performing the logically expected operation?

Reply:
@dekimir Fixed it the way you described (explicitly setting the `_row_ranges`) in the amended commit. 

I think I have now addressed all the comments?